chirpstack/src/maccommand/mod.rs

```diff
@@ -126,6 +133,156 @@ pub async fn handle_uplink<'a>(
     Ok((out, must_respond_with_downlink))
 }
 
+// Checks for MAC commands which the device is unresponsive towards.
+pub async fn check_nonresponsive_mac_commands(
+    uplink_frame_set: &UplinkFrameSet,
+    app: &application::Application,
+    dev: &device::Device,
+    di: &integration_pb::DeviceInfo,
+    ds: &mut internal::DeviceSession,
+    is_relay: bool,
+) {
+    // A list of MAC commands that the device is expected to respond to, if pending.
+    let mut cids: Vec<lrwn::CID> = Vec::from([
+                                       lrwn::CID::NewChannelReq,
+                                       lrwn::CID::LinkADRReq,
+                                       lrwn::CID::RejoinParamSetupReq,
+                                       lrwn::CID::PingSlotChannelReq,
+                                       lrwn::CID::RxParamSetupReq,
+                                       lrwn::CID::RxTimingSetupReq,
+                                       lrwn::CID::TxParamSetupReq
+                                   ]);
+    // A list of MAC commands that the relay is expected to respond to, if pending.
+    let relay_cids: Vec<lrwn::CID> = Vec::from([
+                                         lrwn::CID::UpdateUplinkListReq,
+                                         lrwn::CID::CtrlUplinkListReq,
+                                         lrwn::CID::ConfigureFwdLimitReq,
+                                         lrwn::CID::FilterListReq,
+                                         lrwn::CID::RelayConfReq,
+                                         lrwn::CID::EndDeviceConfReq
+                                     ]);
+
+    if is_relay {
+        cids.extend(relay_cids);
+    }
+
+    do_check_nonresponsive_mac_commands(uplink_frame_set, app, dev, di, ds, cids).await;
+}
+
+async fn do_check_nonresponsive_mac_commands(
+    uplink_frame_set: &UplinkFrameSet,
+    app: &application::Application,
+    dev: &device::Device,
+    di: &integration_pb::DeviceInfo,
+    ds: &mut internal::DeviceSession,
+    cids: Vec<lrwn::CID>,
+) {
+    let conf = config::get();
+    if conf.network.mac_commands_disabled {
+        return;
+    }
+    let max_unack_count = conf.network.max_mac_command_unack_count;
+    let dev_eui = dev.dev_eui;
+
+    let ts: DateTime<Utc> =
+        helpers::get_rx_timestamp(&uplink_frame_set.rx_info_set).into();
+
+    for cid in cids {
+        let unack_count = get_maccommand_unack_count(cid, ds);
+        if unack_count <= max_unack_count {
+            let last_attempt = unack_count == max_unack_count;
+            if check_nonresponsive_mac_command(&dev_eui, cid, ds).await {
```

Comment:
If I understand the code correctly, this will query the Redis database for each `cid` in `cids`, which is going to hurt the overall performance.

Reply:
Yes, it does. I was scared of introducing complexity with my lack of skills, so I settled for a fixed list of CIDs... which is inefficient. If this is a dealbreaker, I'll try to come up with a more agreeable solution.